### DISCUSSION TITLE
How to compat tokio and smol?

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/tokio#4575] [Q&A]
I write my binary crate with `tokio`, and now I want to run my program on ESP32 (a small IoT chip), but it only supports `smol`.

I use `tokio::spawn`, `tokio::sync::RwLock`, `tokio::sync::broadcast`, `tokio::io::{AsyncRead, AsyncWrite}`, `tokio_util::codec::LengthDelimitedCodec`, `tokio::select!` in my program.

Is there some simple way to compat tokio and smol?

**Comment by Darksonn** [comment]
It sounds weird that it would only support smol? I don't see why a chip type would be tied to an async runtime.

But no, there isn't any way to compat them. You can run both runtimes at the same time though.

**Reply by memoryruins** [reply]
> It sounds weird that it would only support smol?

It appears it's not that they only support `smol`, but rather `mio` requires `epoll`/`kqueue` while ESP-IDF's libc only has `poll`/`select`.

https://github.com/ivmarkov/rust-esp32-std-demo/issues/9#issuecomment-899980969

**Reply by lz1998** [reply]
> > It sounds weird that it would only support smol?
> 
> It appears it's not that they only support `smol`, but rather `mio` requires `epoll`/`kqueue` while ESP-IDF's libc only has `poll`/`select`.
> 
> [ivmarkov/rust-esp32-std-demo#9 (comment)](https://github.com/ivmarkov/rust-esp32-std-demo/issues/9#issuecomment-899980969)

Is it possible to patch `tokio` in the same way? I don't want to maintain two repository.

**Reply by memoryruins** [reply]
Based on https://github.com/tokio-rs/mio/issues/1472#issuecomment-796816730 and https://github.com/tokio-rs/mio/issues/1152#issuecomment-991892888, my impression is that it could be possible to add support for it in `mio`, but it would need someone willing to maintain the support.

Another option could be to request support from espressif for something like `epoll` in ESP-IDF, though I'm unsure where would be the best place for that (whether espressif's forum, ESP-IDF's issue tracker, or somewhere else). Perhaps the [`esp-rs` matrix](https://matrix.to/#/#esp-rs:matrix.org) could give some advice.

**Comment by jasta** [comment]
It is now possible to use tokio directly with the esp32 family of devices: https://github.com/jasta/esp32-tokio-demo
